cg/meta/workflow/microsalt/quality_controller/utils.py

```diff
@@ -0,0 +1,142 @@
+from pathlib import Path
+
+from cg.constants.constants import MicrosaltAppTags, MicrosaltQC
+from cg.meta.workflow.microsalt.constants import QUALITY_REPORT_FILE_NAME
+from cg.meta.workflow.microsalt.metrics_parser.models import SampleMetrics
+from cg.meta.workflow.microsalt.quality_controller.models import (
+    CaseQualityResult,
+    SampleQualityResult,
+)
+from cg.meta.workflow.microsalt.quality_controller.result_logger import ResultLogger
+from cg.models.orders.sample_base import ControlEnum
+from cg.store.models import Sample
+
+
+def is_valid_total_reads(reads: int, target_reads: int) -> bool:
+    return reads > target_reads * MicrosaltQC.TARGET_READS_FAIL_THRESHOLD
+
+
+def is_valid_total_reads_for_negative_control(reads: int, target_reads: int) -> bool:
+    return reads < target_reads * MicrosaltQC.NEGATIVE_CONTROL_READS_THRESHOLD
+
+
+def is_valid_mapping_rate(mapping_rate: float) -> bool:
+    return mapping_rate > MicrosaltQC.MAPPED_RATE_THRESHOLD
+
+
+def is_valid_duplication_rate(duplication_rate: float) -> bool:
+    return duplication_rate < MicrosaltQC.DUPLICATION_RATE_THRESHOLD
+
+
+def is_valid_median_insert_size(insert_size: int) -> bool:
+    return insert_size > MicrosaltQC.INSERT_SIZE_THRESHOLD
+
+
+def is_valid_average_coverage(average_coverage: float) -> bool:
+    return average_coverage > MicrosaltQC.AVERAGE_COVERAGE_THRESHOLD
+
+
+def is_valid_10x_coverage(coverage_10x: float) -> bool:
+    return coverage_10x > MicrosaltQC.COVERAGE_10X_THRESHOLD
+
+
+def has_valid_mapping_rate(metrics: SampleMetrics) -> bool:
+    mapped_rate: float | None = metrics.microsalt_samtools_stats.mapped_rate
+    return is_valid_mapping_rate(mapped_rate) if mapped_rate else False
+
+
+def has_valid_duplication_rate(metrics: SampleMetrics) -> bool:
+    duplication_rate: float | None = metrics.picard_markduplicate.duplication_rate
+    return is_valid_duplication_rate(duplication_rate) if duplication_rate else False
+
+
+def has_valid_median_insert_size(metrics: SampleMetrics) -> bool:
+    insert_size: int | None = metrics.picard_markduplicate.insert_size
+    return is_valid_median_insert_size(insert_size) if insert_size else False
+
+
+def has_valid_average_coverage(metrics: SampleMetrics) -> bool:
+    coverage: float | None = metrics.microsalt_samtools_stats.average_coverage
+    return is_valid_average_coverage(coverage) if coverage else False
+
+
+def has_valid_10x_coverage(metrics: SampleMetrics) -> bool:
+    coverage_10x: float | None = metrics.microsalt_samtools_stats.coverage_10x
+    return is_valid_10x_coverage(coverage_10x) if coverage_10x else False
+
+
+def get_negative_control_result(results: list[SampleQualityResult]) -> SampleQualityResult:
+    for result in results:
+        if result.is_control:
+            return result
+    raise ValueError("No negative control found")
```

Comment:
This is incorrect. It is not required that an order includes a negative control. Most of them do, but it is not required.

Reply:
Alternatively, you could first check if the case includes a negative control from statusdb, and if it does and it is not here you could fail it. But that seems like a workaround - I just think you can remove the raise and instead log or not that there is no negative control in the report.